Modules/_sqlite/connection.c

```diff
@@ -321,10 +378,33 @@ free_callback_contexts(pysqlite_Connection *self)
     set_callback_context(&self->authorizer_ctx, NULL);
 }
 
+static void
+remove_callbacks(sqlite3 *db)
+{
+#ifdef HAVE_TRACE_V2
+    sqlite3_trace_v2(db, SQLITE_TRACE_STMT, 0, 0);
+#else
+    sqlite3_trace(db, 0, (void*)0);
+#endif
+    sqlite3_progress_handler(db, 0, 0, (void *)0);
+    (void)sqlite3_set_authorizer(db, NULL, NULL);
+}
+
 static void
 connection_close(pysqlite_Connection *self)
 {
     if (self->db) {
+        if (self->autocommit == AUTOCOMMIT_DISABLED &&
+            !sqlite3_get_autocommit(self->db))
+        {
+            /* If close is implicitly called as a result of interpreter
+             * tear-down, we must not call back into Python. */
+            if (_Py_IsFinalizing()) {
+                remove_callbacks(self->db);
+            }
+            (void)connection_exec_stmt(self, "ROLLBACK");
```

Comment:
> But is an implicit rollback necessary since uncommitted changes won’t be saved anyway?

The alternative is to rely on SQLite to _not_ commit the pending transaction on close. We can probably rely on SQLite keeping the current behaviour, but it means we rely on the underlying implementation in order to fulfil the PEP 249 contract. I prefer to be explicit about that contract.

> And if we still do an implicit rollback, why don’t we do it in the new auto-commit mode as well [...]

See line 397 in `Modules/_sqlite/connection.c` in this PR; the implicit rollback only happens if the new autocommit mode is used. The new PEP 249 specification specifies two modes: `autocommit is True` and `autocommit is False`. The latter is what MAL calls "transactional mode" (I cannot recall his exact wording), where a transaction is _always open_, and the user is expected to use `.commit()` and `.rollback()` explicitly. In the former mode, aka "autocommit mode", we do not emit any implicit statement, hence no implicit rollback at close.

> A transaction might have been initiated (explicitly with a BEGIN statement) [...]

This has been discussed on this issue (and I believe also on the db-sig ml). It was decided that we do not try to guard the user against shooting themselves in the foot; if they want to use explicit transaction handling, they are completely on their own. If we were to try and catch such scenarios, we'd have to introduce a lot of complicated code (for example an SQL parser), which would be hard (or close to impossible) to test; the added complexity would not be worth it. How could we even make sure we'd be able to guard against future features in SQLite? They introduce new keywords and new behaviour all the time. The best thing to do is to keep the code simple and just follow PEP 249. If the user breaks the PEP 249 contract, they are on their own.

Reply:
> The alternative is to rely on SQLite to not commit the pending transaction on close. We can probably rely on SQLite keeping the current behaviour, but it means we rely on the underlying implementation in order to fulfil the PEP 249 contract. I prefer to be explicit about that contract.

> In the former mode, aka "autocommit mode", we do not emit any implicit statement, hence no implicit rollback at close.

That makes sense.